MdeModulePkg/Core/Pei/Memory/MemoryBuckets.c

```diff
@@ -0,0 +1,470 @@
+  @param[in] DataSize      The size of the data in the MEMORY_TYPE_INFORMATION
+                           hob.
+  @param[in] MemInfo       Pointer the the MEMORY_TYPE_INFORMATION object
+                           with the relevant memory bucket sizes.
+  @param[in] MemoryType    The type of memory we are interested in.
+
+  @retval    Returns the number of pages to use for the bucket in the memory
+             memory type inputted.
+**/
+UINT32
+EFIAPI
+GetBucketSizeFromMemoryInfoHob (
+  IN PEI_CORE_INSTANCE     *PrivateData,
+  IN UINTN                  DataSize,
+  IN EFI_MEMORY_TYPE_INFORMATION *MemInfo,
+  IN EFI_MEMORY_TYPE       MemoryType
+  )
+{
+  UINTN Index;
+  for (Index = 0; Index < DataSize / sizeof (EFI_MEMORY_TYPE_INFORMATION); Index++) {
+    if (MemInfo[Index].Type == (UINT32)MemoryType) {
+      return MemInfo[Index].NumberOfPages;
+    }
+  }
+  DEBUG ((DEBUG_ERROR, "[%a] - The memory time wasn't found in the MEMORY_TYPE_INFORMATION hob!.\n", __FUNCTION__));
+  return 0;
+}
+
+/**
+  This function initialized the PEI memory buckets.  This should be called
+  when the first memory type being tracked in these buckets is allocated.
+
+      +-----------------------------+  <-- Top of free memory
+      |  Memory Bucket Allocation 1 |
+      +-----------------------------+  <-- Top of free memory + 1st non-zero memory bucket pcd
+      |  Memory Bucket Allocation 2 |
+      +-----------------------------+  <-- Top of free memory + 1st and 2nd non-zero memory bucket pcds
+      |  Memory Bucket Allocation 3 |
+      +-----------------------------+  <-- Top of free memory + 1st, 2nd, 3rd non-zero memory bucket pcds
+      |   . . . . . . . . . . . .   |
+      +-----------------------------+  <-- Top of free memory + 1-n non-zero memory bucket pcds
+      |         FREE MEMORY         |      For PEI allocations all new non bucket free memory starts here.
+      +-----------------------------+
+
+  NOTE: The memory buckets can be created in Pre-mem PEI if runtime pages are
+  allocated in pre-mem PEI.
+
+  @param[in] PrivateData   Pointer to PeiCore's private data structure.
+  @param[in] StartingAddress   The starting address of the memory buckets.
+                               All of them are contiguous.
+**/
+VOID
+EFIAPI
+InitializeMemoryBuckets (
+  IN PEI_CORE_INSTANCE     *PrivateData,
+  IN EFI_PHYSICAL_ADDRESS  StartingAddress
+  )
+{
+  UINTN   Index;
+  UINT64  AddressAdjustment;
+
+  AddressAdjustment = 0;
+
+  for (Index = 0; Index < EfiMaxMemoryType+1; Index++) {
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].BaseAddress = 0;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].MaximumAddress = MAX_ALLOC_ADDRESS;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].CurrentNumberOfPages = 0;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].NumberOfPages = 0;
+    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].InformationIndex = mMemoryTypes[Index];

```

Comment:
Looks to me like this should just be `Index` based on usage I see elsewhere.

```suggestion
    PrivateData->PeiMemoryBuckets.RuntimeBuckets[Index].InformationIndex = Index;

```